### PR TITLE
Mark `nonzero` parameters experimental in sparse min/max

### LINKED ISSUE #3535: Difference from SciPy about sparse min/max `sum_duplicates` parameter
I'm not sure its background, but currently, sparse min/max have `sum_duplicates` parameter and its default value is `False`. However, their SciPy counterparts don't have the parameter and always sum duplicate entries as `sum_duplicates=True`.

It would be better to change its default value to `False`.

### PR DESCRIPTION
**Pull request by takagi**
~~Close #3535.~~

~~Currently, sparse min/max/argmin/argmax have `sum_duplicates` and `nonzero` parameters. However, the SciPy counterparts do not have them and always conduct as `sum_duplicates=True` and `nonzero=False`. This PR removes the parameters from the methods to make compliance with SciPy.~~

Merge #3676 first.

Currently, sparse min/max have `nonzero` parameter that indicates that the operations should take care of explicitly stored elements only in a sparse matrix. However, the parameter is now proposed in https://github.com/scipy/scipy/pull/11899 and is not merged to the SciPy master nor released in its stable release. This PR makes it a keyword-only parameter and marks as experimental for future possible changes.

### CODE REVIEW

**Comment by dloney** [comment]
I would say toggling the default behavior from False to True to be consistent with scipy would be good. The reason I introduced the flag is as a performance enhancement to skip the sum_duplicates operation if the user knows their matrix is canonical. Changing the default flag will make it conform with the default scipy behavior.

For #3583, there's an open pull request on scipy that will implement the explicit nonzero features. I wouldn't remove the functionality from cupy for the sake of standardizing to scipy since it's a useful feature when explicit zeros are included in the sparse matrix. The default behavior for the non_zero flag is the same as standard scipy. The scipy pull request uses the keyword "explicit" rather than "nonzero" which might be a good change to make.

**Comment by leofang** [comment]
I think it's best to stay complaint with the current SciPy, not the SciPy to be, as it could take months to get the PR merged in the upstream. This would also ensure compliance with CuPy's own dev policy.

**Comment by leofang** [comment]
ref: https://github.com/scipy/scipy/pull/11899 (repost here for easier reference)

**Comment by dloney** [comment]
I like the suggestion from @grlee77 in #2711 to keep the functionality but shift it to use keywords. That would keep the same behavior as scipy as well as the explicit zero functionality. 

It's important to keep the functionality available since it provides an order of magnitude speedup when there are explicit zeros in the the sparse matrix.

**Comment by takagi** [comment]
@dloney, I agree with you that the explicit zero functionality is useful. Now I think of when and how to introduce it to CuPy's **stable** version.

We have our [API Compatibility Policy](https://docs-cupy.chainer.org/en/stable/compatibility.html), where we can make changes that break the backward compatibility only in major updates. Incompatible changes here include those as small as modifying parameter names. The next major update (8.0.0) is planned on July 30, which is soon and the upstream won't be fixed before the date.

I suppose we have the following options:
- Introducing to 8.0.0 with making it a keyword-only parameter (@grlee77 's suggestion)
- Introducing to 8.0.0 with making it keyword-only and marking as [cupy.util.experimental](https://github.com/cupy/cupy/blob/92ab7b427691dd0dd7172b562d814d640a54eb7c/cupy/util.pyx#LC119) until SciPy supports it
- Removing from 8.0.0 and introducing to (maybe) 9.0.0 after SciPy supports it (@leofang 's suggestion)

I prefer the third option as @leofang says for making things cleaner, but I think the second is acceptable. How do you think?

**Comment by leofang** [comment]
Marking it experimental would not be bad as long as it helps!

**Comment by dloney** [comment]
I'd be happy marking it as experimental. The core functionality probably shouldn't have the experimental flag applied to it though since it replicates what's already available in scipy. 

I'm not familiar with the cupy implementation of the experimental flag . Would we be able to invoke the experimental flag only when the keyword argument is enabled?  Looking it over quick, it looks like it would be adding an if statement with additional import were the keyword enabled.

**Comment by grlee77** [comment]
> I prefer the third option as @leofang says for making things cleaner, but I think the second is acceptable. How do you think?

I am not using it personally, but given that it has already been implemented and it seems at least some are finding it useful, I would lean toward leaving it in, but marking it as experimental

**Comment by takagi** [comment]
> Would we be able to invoke the experimental flag only when the keyword argument is enabled?

Yes, we can call the flag only when the keyword argument is used. The flag is implemented as a function instead of a decorator for such a case.

Okay, then I'll change this PR as such. Thanks, guys!

**Comment by takagi** [comment]
I fixed this PR to make the parameters keyword-only and to mark them experimental.

Now please let me make sure about the `sum_duplicates` parameter, @dloney. `sum_duplicates` method returns immediately when a sparse matrix is already canonical. Do we still need this?

https://github.com/cupy/cupy/blob/725c350884142c85c78df13a111f29b7c508540e/cupyx/scipy/sparse/compressed.py#L708-L709

**Comment by dloney** [comment]
No, I don't believe so especially given the proposed scipy format.

**Comment by dloney** [comment]
No, I don't believe so especially given the proposed scipy format. Do we also want to change the parameter name from nonzero to explcit to match scipy?

**Comment by takagi** [comment]
I've extracted removing `sum_duplicate` parameter to #3676 to make the changes in PRs smaller.

File: cupyx/scipy/sparse/data.py [hunk 1]
```diff
@@ -298,11 +297,10 @@ def min(self, axis=None, out=None, sum_duplicates=False, nonzero=False):
...
                 the default value, as this argument is not used.
-            sum_duplicates (bool): Flag to indicate that duplicate elements
-                should be combined prior to the operation
-            nonzero (bool): Return the minimum nonzero value and ignore all
-                zero entries. If the dimension has no nonzero values, a zero is
-                then returned to indicate that it is the only available value.
+            explicit (bool): Return the minimum value explicitly specified and
+                ignore all implicit zero entries. If the dimension has no
+                explicit values, a zero is then returned to indicate that it is
+                the only implicit value.
```
Comment (emcastillo) [inline cupyx/scipy/sparse/data.py:303]: Let's add a note about explicit being experimental

Comment (takagi) [reply cupyx/scipy/sparse/data.py]: Added!

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**